### PR TITLE
fs: bring back macos-specific copyfile(3)

### PR DESCRIPTION
**Pull request by mischnic**
An updated version of #2578, but using `faccessat` and removing the unecessary version checks (addressing the feedback from @bnoordhuis)

Closes #2578

See the linked PR for the previous discussion.

I'm not entirely sure if I'm using `faccessat` correctly here (and which flags should be passed).

### CODE REVIEW

**Comment by cjihrig** [comment]
The `fs_copyfile` test is failing on macOS in the CI. Seems relevant.

**Comment by mischnic** [comment]
Indeed, but strangely enough that test passes for me locally on macOS 12.2.1 ...

**Comment by mischnic** [comment]
Apparently, `copyfile` truncates the file on macOS 10.15 if source = destination, but not on macOS 12. I guess I'll have to add  a check just like the one in the non-macOS branch: https://github.com/libuv/libuv/blob/1a91b51976a1adc6972081faa78b6b70022254d3/src/unix/fs.c#L1324-L1328

**Comment by wojtekmaj** [comment]
Ping

File: src/unix/fs.c [hunk 1]
```diff
@@ -1264,6 +1265,32 @@ static ssize_t uv__fs_write(uv_fs_t* req) {
...
+  /* On macOS, use the native copyfile(3). */
+  copyfile_flags_t flags;
+
+  /* Don't overwrite the destination if its permissions disallow it. */
+  if (faccessat(AT_FDCWD, req->new_path, R_OK | W_OK, AT_EACCESS)) {
+    if (errno != ENOENT)
+      return UV__ERR(errno);
+  }
+
+  flags = COPYFILE_ALL;
```
Comment (kwasimensah) [inline src/unix/fs.c:1278]: Should this also specifcy COPYFILE_ACL to make sure permissions get copied over (and a new bug won't be opened about permissions not being respected?)

**Comment by sanmai-NL** [comment]
@bnoordhuis ping

**Comment by sanmai-NL** [comment]
The suggested patch hadn't been applied and now it has.

How to check the failing test's output?

**Comment by mischnic** [comment]
On CI, there's currently this failing test:
```
1: Test command: /Users/runner/work/libuv/libuv/build/uv_run_tests
....
1: not ok 57 - fs_event_watch_dir
1: # timeout
1: # Output from process `fs_event_watch_dir`: (no output)
```

I can't reproduce that locally, so it might be a flakey test?

<br>

The problem in https://github.com/libuv/libuv/pull/3654#issuecomment-1172215332 isn't actually covered by CI anymore because it now runs on macOS 11 (and 10.15 is technically older than the supported [`Current and previous macOS release`](https://github.com/libuv/libuv/blob/a73725579107e374eb8358a3999d03b822df6c89/SUPPORTED_PLATFORMS.md)). But the error was
```
$ build/uv_run_tests
...
not ok 45 - fs_copyfile
# exit code 6
# Output from process `fs_copyfile`:
# Assertion failed in libuv/test/test-fs-copyfile.c on line 131: `req.statbuf.st_size == 12` (0 == 12)
```

**Comment by kylemh** [comment]
the github actions seem to [run on macos-11](https://github.com/libuv/libuv/blob/v1.x/.github/workflows/CI-unix.yml#L36), maybe that explains the discrepancy?

**Comment by bnoordhuis** [comment]
@mischnic can you rebase? I wanted to re-run the CI but automatic rebase fails. Thanks.

**Review by bnoordhuis — APPROVED** [review]
Tests pass, yay. Thanks @mischnic, I'll go ahead and merge it. In the nick of time for the next release. :-)

**Comment by karlhorky** [comment]
Nice, thanks @mischnic and others who helped with this!

> In the nick of time for the next release

Really cool, I'm guessing `1.45.0`? (minor?) Is there anywhere public with the release schedule for `libuv`? I looked at [the Milestones](https://github.com/libuv/libuv/milestones), but I guess this will not be a part of `2.0.0`... 🤔

**Comment by santigimeno** [comment]
Yes, `1.45.0`. There's no regular release schedule. Usually we cut a release when we think there's enough new features/fixes. (or in the event of some major regression)

**Comment by zkochan** [comment]
It would make a big difference for pnpm (maybe for Yarn too). We currently have to use hard links on mac and copies are much better. It would be very impactful for many devs.

**Comment by clason** [comment]
At Neovim (see the linked PR), we are seeing new test failures involving copying and symbolic links.

I suspect (but haven't tested yet) that this is due to the flags set -- could it be that you are missing a `AT_SYMLINK_NOFOLLOW{_ANY}` flag in `faccessat` here?

File: src/unix/fs.c [hunk 2]
```diff
@@ -1240,6 +1241,32 @@ static ssize_t uv__fs_write(uv_fs_t* req) {
...
+
+  if (req->flags & UV_FS_COPYFILE_FICLONE)
+    flags |= COPYFILE_CLONE;
```
Comment (clason) [inline src/unix/fs.c:1257]: Specifically, it's this flag that leads to wrong behavior -- it creates a symlink, where the previous UV implementation copied the file itself.

**Comment by clason** [comment]
From `man copyfile`:
> Since this flag implies COPYFILE_NOFOLLOW_SRC, symbolic links themselves will be cloned instead of their targets.

I believe that is an unintended change in behavior? Maybe the flags here should be explicitly 
```
COPYFILE_EXCL | COPYFILE_ACL | COPYFILE_STAT | COPYFILE_XATTR | COPYFILE_DATA
```
(and similarly for `COPYFILE_CLONE_FORCE`).

**Pull request merged**